Task_1/task1_file_parser.py

```diff
@@ -0,0 +1,49 @@
+#!/bin/python3
+import os
+import sys
+from datetime import datetime
+
+#file_path = sys.argv[1]
+file_path = 'C:\PythonMentorship\Task_1\input\log_data\log_data.txt'
+
+def group_by_date_and_IP (record_list):
+    grouped_by_date = {}
+    for record in record_list:
+        if record["Datetime"].date() not in grouped_by_date:
+            grouped_by_date[record["Datetime"].date()] = {}
+        grouped_by_IP = grouped_by_date[record["Datetime"].date()]
+        if record["IP"] not in grouped_by_IP:
+            grouped_by_IP[record["IP"]] = record["Size"]
+        else:
+            grouped_by_IP[record["IP"]] += record["Size"]
+    return grouped_by_date
+
+
+def print_top_downloader(grouped_by_date):
+    print('IP address of the top downloader: ')
+    for date, ip in grouped_by_date.items():
+        top_downloader = max(ip, key = ip.get)
+        print(f'{date} - {top_downloader}')
+
+
+def get_least_busy_hours(record_list):
+    list_of_hours = dict.fromkeys(range(00,24),0)
+    for record  in record_list:
+        list_of_hours[record['Datetime'].hour] += 1
+    least_busy_hour = (min(list_of_hours,key = list_of_hours.get), list_of_hours[min(list_of_hours,key = list_of_hours.get)])
+    return least_busy_hour
+
+with open(file_path,'r') as log_data:
+    log_data_record = []
+    for line in log_data:
+        line = line.replace('\n','')
+        str = line.split(' ')

```

Comment:
Please note that you reassign the built-in str to a different value which can have dangerous side effects.
You can spot that since the editor highlighted the "str" as a keyword. Mind that Python will technically allow you to write the following (this will not raise exceptions):
True = 0
So please use a different var name (e.g. "parts") so that there is no clashing with the built-in name.

Reply:
Done